Refactor quad loop

charleskawczynski opened the pull request:
This is a step towards putting the quadrature loop into the form: `quad(f, m, cov, dist)`, which evaluates an integral `int pdf(x; m, cov) .* f(x) dx`. Even just with this PR, I think the physics are more decoupled from the integration. There is a TODO added, which I'm not yet sure how to address:

```julia
σ_q_lim = -qt_mean / (sqrt2 * env_thermo_quad.a[1]) # TODO: is this correct?
```

For now,t there's an inference issue which requires us to type cast in `compute_cloud_fraction`, but we should be able to revert it once we dig into what's wrong there.

szy21 commented on src/prognostic_equations/cloud_fraction.jl:
```diff
@@ -96,67 +88,53 @@ function quad_loop(env_thermo_quad::SGSQuadrature, vars, thermo_params)
-
-        for idx in 1:env_len
-            outer_env[idx] += inner_env[idx] * weights[m_q] * sqpi_inv
+            inner_env = inner_env ⊞ f(χ[m_h]) ⊠ weights[m_h] ⊠ FT(1 / sqrt(π))
```
Why do we need RecursiveApply?

charleskawczynski replied:
This will allow us to compute quadrature over arbitrary variables in one pass (e.g., a vector of scalars, a NamedTuple of scalars, structs etc.)

szy21 commented:
It might be good if @tapios takes a look at this as well. If I understand Tapio's comments correctly, what we want is a function f(qt_hat, thetali_hat) rather than f(χ). But qt_hat and thetali_hat are just a function of χ and mean values so maybe this is ok as well? In any case this seems to be a step towards decoupling the pdf and the physical function, thanks!

tapios commented:
It's going in the right direction but can go further. I want to decouple all thermodynamics from the quadrature. That is, write a function that evaluates the bivariate quadrature `int pdf(x_1, x_2, parameters; m, cov) .* f(x_1, x_2) dx_1 dx_2` (my previous x was supposed to be a 2-vector). Then evaluate this with a function such as `f(q_t, theta_li) = heaviside (q_t - q_v^*(q_t, theta_li, pars))`, where `pars`, this case, would include density. That way, the same quadrature function could be used for any combination of thermodynamic variables we choose, and no explicit thermodynamic evaluations (other than those necessary to evaluate `f`) would be needed in the quadrature function itself.

charleskawczynski commented:
> It's going in the right direction but can go further. I want to decouple all thermodynamics from the quadrature. That is, write a function that evaluates the bivariate quadrature `int pdf(x_1, x_2, parameters; m, cov) .* f(x_1, x_2) dx_1 dx_2` (my previous x was supposed to be a 2-vector). Then evaluate this with a function such as `f(q_t, theta_li) = heaviside (q_t - q_v^*(q_t, theta_li, pars))`, where `pars`, this case, would include density. That way, the same quadrature function could be used for any combination of thermodynamic variables we choose, and no explicit thermodynamic evaluations (other than those necessary to evaluate `f`) would be needed in the quadrature function itself.

Agreed, I thought that this would be a good intermediate step, I think we can make the remaining changes in a follow up PR.

charleskawczynski commented:
Ok, inference was really problematic in that function because the compiler was boxing many things (due to redefining variables and the compiler didn't seem to like defining different closures inside union split branches). So, I removed the log normal distribution (for now), changed a few variable names, and added some type casting to fix type instabilities (`eps(Float32) isa Float64`!)

charleskawczynski commented:
@szy21, can you take a second look? This may make this function significantly faster too (it was previously _very_ expensive, about as expensive as `step!`!)

charleskawczynski commented:
In case I need to re-review, this is what I was using to debug via JET:

```julia
import Revise
import ClimaCore
import FastGaussQuadrature
import Thermodynamics as TD
import CLIMAParameters as CP
if !(@isdefined TU)
    include(
        joinpath(pkgdir(ClimaCore), "test", "TestUtilities", "TestUtilities.jl"),
    );
end
import .TestUtilities as TU;
space = TU.CenterExtrudedFiniteDifferenceSpace(Float64);
f = fill((; ᶜts = zero(TD.PhaseEquil{Float64}), ᶜp = Float64(101325)), space);
@. f.ᶜts = TD.PhaseEquil_ρTq(thermo_params, 1.0, 300.0, 0.001);
@. f.ᶜp = TD.air_pressure(thermo_params, f.ᶜts);


include("src/cache/precomputed_quantities.jl")
include("src/prognostic_equations/cloud_fraction.jl")

aliases = string.(fieldnames(TD.Parameters.ThermodynamicsParameters));
toml_dict = CP.create_toml_dict(Float64; dict_type = "alias");
pairs = CP.get_parameter_values!(toml_dict, aliases, "Thermodynamics");
thermo_params = TD.Parameters.ThermodynamicsParameters{Float64}(; pairs...);
Base.broadcastable(x::SGSQuadrature) = tuple(x)
env_thermo_quad = SGSQuadrature(Float64);
get_cloud_fraction(thermo_params, env_thermo_quad, f.ᶜp, f.ᶜts);

using JET
include("src/prognostic_equations/cloud_fraction.jl"); JET.@test_opt get_cloud_fraction(thermo_params, env_thermo_quad, f.ᶜp, f.ᶜts);
```

charleskawczynski commented:
bors r+

charleskawczynski commented:
bors r+

charleskawczynski commented:
bors r+

szy21 commented:
bors r+

szy21 commented:
bors r+

szy21 commented:
bors r+
